ambry-router/src/main/java/com.github.ambry.router/CryptoJob.java

```diff
@@ -0,0 +1,230 @@
+     * @param kms the {@link KeyManagementService} to fetch keys from
+     */
+    public void doOperation(CryptoService cryptoService, KeyManagementService kms) {
+      try {
+        ByteBuffer encryptedContent = cryptoService.encrypt(toEncrypt, perBlobKey);
+        Object containerKey = kms.getKey(blobId.getAccountId(), blobId.getContainerId());
+        ByteBuffer encryptedKey = cryptoService.encryptKey(perBlobKey, containerKey);
+        callback.onCompletion(new EncryptJobResult(blobId, encryptedKey, encryptedContent), null);
+      } catch (Exception e) {
+        callback.onCompletion(new EncryptJobResult(blobId, null, null), e);
+      }
+    }
+
+    /**
+     * Close the job with the given {@code gse}
+     * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callback for the job
+     */
+    public void closeJob(GeneralSecurityException gse) {
+      callback.onCompletion(new EncryptJobResult(blobId, null, null), gse);
+    }
+  }
+
+  /**
+   * Class representing an decrypt Job.
+   */
+  class DecryptJob implements CryptoJob {
+    private final BlobId blobId;
+    private final ByteBuffer enryptedContent;
+    private final ByteBuffer encryptedKey;
+    private final DecryptCallback callback;
+
+    /**
+     * Instantiates {@link DecryptJob} with {@link BlobId}, key to be decrypted, content to be decrypted and the
+     * {@link DecryptCallback}
+     * @param blobId the {@link BlobId} for which decryption is requested
+     * @param encryptedKey encrypted per blob key
+     * @param encryptedContent encrypted blob content
+     * @param callback {@link DecryptCallback} to be invoked on completion
+     */
+    DecryptJob(BlobId blobId, ByteBuffer encryptedKey, ByteBuffer encryptedContent, DecryptCallback callback) {
+      this.blobId = blobId;
+      this.enryptedContent = encryptedContent;
+      this.encryptedKey = encryptedKey;
+      this.callback = callback;
+    }
+
+    /**
+     * Steps to be performed on decryption
+     * 1. Fetch ContainerKey from kms for the given blob
+     * 2. Decrypt encryptedKey using containerKey to obtain perBlobKey
+     * 3. Decrypt encryptedContent using perBlobKey
+     * 4. Invoke callback with the decryptedContent
+     * @param cryptoService the {@link CryptoService} to assist in encrypting and decrypting
+     * @param kms the {@link KeyManagementService} to fetch keys from
+     */
+    public void doOperation(CryptoService cryptoService, KeyManagementService kms) {
+      try {
+        Object containerKey = kms.getKey(blobId.getAccountId(), blobId.getContainerId());
+        Object perBlobKey = cryptoService.decryptKey(encryptedKey, containerKey);
+        ByteBuffer decryptedContent = cryptoService.decrypt(enryptedContent, perBlobKey);
+        callback.onCompletion(new DecryptJobResult(blobId, decryptedContent), null);
+      } catch (Exception e) {
+        callback.onCompletion(new DecryptJobResult(blobId, null), e);
```

Comment:
Is the result required on exceptions?

Reply:
As I mentioned above, or handling error scenarios, this might come in handy